server/lorax_server/models/flash_phi.py

```diff
@@ -0,0 +1,178 @@
+import torch
+import torch.distributed
+
+from loguru import logger
+from opentelemetry import trace
+from transformers import AutoTokenizer
+from typing import Dict, List, Optional, Tuple
+
+from lorax_server.models import FlashCausalLM
+from lorax_server.models.custom_modeling.flash_phi_modeling import (
+    ATTN_WQKV,
+    ATTN_OUT_PROJ,
+    MLP_FC1,
+    MLP_FC2,
+    FlashPhiForCausalLM,
+    PhiConfig,
+)
+from lorax_server.utils import (
+    create_merged_weight_files,
+    initialize_torch_distributed,
+    weight_files,
+    Weights,
+)
+from lorax_server.utils.adapter import BASE_MODEL_ADAPTER_ID
+from lorax_server.utils.lora import LM_HEAD
+from lorax_server.utils.weights import shard_on_dim
+
+tracer = trace.get_tracer(__name__)
+
+
+ADAPTER_LAYERS = [ATTN_WQKV, ATTN_OUT_PROJ, MLP_FC1, MLP_FC2, LM_HEAD]
+ROW_PARALLEL = {ATTN_OUT_PROJ, MLP_FC2, LM_HEAD}
+
+
+class FlashPhi(FlashCausalLM):
+    def __init__(
+        self,
+        model_id: str,
+        adapter_id: str,
+        adapter_source: str,
+        revision: Optional[str] = None,
+        quantize: Optional[str] = None,
+        dtype: Optional[torch.dtype] = None,
+        trust_remote_code: bool = False,
+    ):
+        self.process_group, rank, world_size = initialize_torch_distributed()
+        if torch.cuda.is_available():
+            device = torch.device(f"cuda:{rank}")
+            dtype = torch.float16 if dtype is None else dtype
+        else:
+            raise NotImplementedError("FlashPhi is only available on GPU")
+
+        tokenizer = AutoTokenizer.from_pretrained(
+            model_id,
+            revision=revision,
+            padding_side="left",
+            truncation_side="left",
+            trust_remote_code=trust_remote_code,
```

Comment:
Since we're adding `trust_remote_code`, do we already install the `einops` library to convert weights when using `microsoft/phi-1.5`?

Reply:
`trust_remote_code` is `False` by default. `einops` is installed already (for Falcon). This current implementation works with the microsoft version of the weights, rather than the changes HF made.